File: go.sdk/src/bma/servicecall/sockcore/messagewriter.go
package sockcore

import (
	sccore "bma/servicecall/core"
	"bytes"
	"io"
)

const (
	HEADER_SIZE = 4
)

var (
	endData  = []byte{0, 0, 0, 0}
	flagData = []byte{2, 0, 0, 0}
)

type MessageWriter struct {
	w    io.Writer
	ew   EncodeWriter
	hbuf []byte
	buf  *bytes.Buffer
}

func NewMessageWriter(w io.Writer) *MessageWriter {
	o := new(MessageWriter)
	o.w = w
	if ew, ok := w.(EncodeWriter); ok {
		o.ew = ew
	}
	o.hbuf = make([]byte, 4)
	return o
}

func (this *MessageWriter) Close() {
	if this.buf != nil {
		this.buf.Reset()
		this.buf = nil
	}
}

func (this *MessageWriter) Write(p []byte) (n int, err error) {
	// fmt.Printf("WRITE <- %v\n", p)
	return this.w.Write(p)
}

func (this *MessageWriter) WriteByte(c byte) error {
	// fmt.Printf("WRITE <- %v\n", c)
	if this.ew != nil {
		return this.ew.WriteByte(c)
	}
	bs := []byte{c}
	_, err := this.w.Write(bs)
	return err
}

func (this *MessageWriter) WriteHeader(mt byte, sz int) error {
	this.hbuf[0] = mt
	this.hbuf[1] = byte(sz >> 16)
	this.hbuf[2] = byte(sz >> 8)
	this.hbuf[3] = byte(sz)
	_, err := this.Write(this.hbuf)
	return err
}

func (this *MessageWriter) WriteEnd() error {
	_, err := this.Write(endData)
	return err
}

func (this *MessageWriter) WriteMessageId(mid int32) error {
	this.WriteHeader(MT_MESSAGE_ID, 4)
	return Coders.FixUint32.DoEncode(this, uint32(mid))
}

func (this *MessageWriter) WriteFlag() error {
	_, err := this.Write(flagData)
	return err
}

func (this *MessageWriter) sbuf() *bytes.Buffer {
	if this.buf == nil {
		this.buf = bytes.NewBuffer([]byte{})
	}
	this.buf.Reset()
	return this.buf
}

func (this *MessageWriter) WriteAddress(s string, m string) error {
	buf := this.sbuf()
	err1 := Coders.LenString.DoEncode(buf, s)
	if err1 != nil {
		return err1
	}
	err2 := Coders.LenString.DoEncode(buf, m)
	if err2 != nil {
		return err2
	}
	err0 := this.WriteHeader(MT_ADDRESS, buf.Len())
	if err0 != nil {
		return err0
	}
	_, errW := this.Write(buf.Bytes())
	if errW != nil {
		return errW
	}
	return nil
}

func (this *MessageWriter) WriteData(n string, val *sccore.Value) error {
	buf := this.sbuf()
	err1 := Coders.LenString.DoEncode(buf, n)
	if err1 != nil {
		return err1
	}
	var v interface{}
	if val != nil {
		v = val.ToValue()
	}
	err2 := Coders.Varinat.Encode(buf, v)
	if err2 != nil {
		return err2
	}
	err0 := this.WriteHeader(MT_DATA, buf.Len())
	if err0 != nil {
		return err0
	}
	_, errW := this.Write(buf.Bytes())
	if errW != nil {
		return errW
	}
	return nil
}

func (this *MessageWriter) WriteContext(n string, val *sccore.Value) error {
	buf := this.sbuf()
	err1 := Coders.LenString.DoEncode(buf, n)
	if err1 != nil {
		return err1
	}
	var v interface{}
	if val != nil {
		v = val.ToValue()
	}
	err2 := Coders.Varinat.Encode(buf, v)
	if err2 != nil {
		return err2
	}
	err0 := this.WriteHeader(MT_CONTEXT, buf.Len())
	if err0 != nil {
		return err0
	}
	_, errW := this.Write(buf.Bytes())
	if errW != nil {
		return errW
	}
	return nil
}

func (this *MessageWriter) WriteAnswer(st int, msg string) error {
	buf := this.sbuf()
	err1 := Coders.Int32.DoEncode(buf, int32(st))
	if err1 != nil {
		return err1
	}
	err2 := Coders.LenString.DoEncode(buf, msg)
	if err2 != nil {
		return err2
	}
	err0 := this.WriteHeader(MT_ANSWER, buf.Len())
	if err0 != nil {
		return err0
	}
	_, errW := this.Write(buf.Bytes())
	if errW != nil {
		return errW
	}
	return nil
}

func (this *MessageWriter) SendRequest(mid int32, s, m string, req *sccore.Request, ctx *sccore.Context) error {
	var err error
	err = this.WriteMessageId(mid)
	if err != nil {
		return err
	}
	err = this.WriteFlag()
	if err != nil {
		return err
	}
	err = this.WriteAddress(s, m)
	if err != nil {
		return err
	}
	if req != nil {
		req.Walk(func(k string, v *sccore.Value) bool {
			err = this.WriteData(k, v)
			return err != nil
		})
		if err != nil {
			return err
		}
	}
	if ctx != nil {
		ctx.Walk(func(k string, v *sccore.Value) bool {
			err = this.WriteContext(k, v)
			return err != nil
		})
		if err != nil {
			return err
		}
	}
	return this.WriteEnd()
}

func (this *MessageWriter) SendAnswer(mid int32, an *sccore.Answer) error {
	err := this.WriteMessageId(mid)
	if err != nil {
		return err
	}
	this.WriteAnswer(an.GetStatus(), an.GetMessage())

	var werr error
	rs := an.GetResult()
	if rs != nil {
		rs.Walk(func(k string, v *sccore.Value) bool {
			werr = this.WriteData(k, v)
			return werr != nil
		})
		if werr != nil {
			return werr
		}
	}

	ctx := an.GetContext()
	if ctx != nil {
		ctx.Walk(func(k string, v *sccore.Value) bool {
			werr = this.WriteContext(k, v)
			return werr != nil
		})
		if werr != nil {
			return werr
		}
	}
	return this.WriteEnd()
}

File: java.sdk/sockcore/src/main/java/bma/servicecall/sockcore/MessageWriter.java
package bma.servicecall.sockcore;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.OutputStream;

import bma.servicecall.core.Answer;
import bma.servicecall.core.AppError;
import bma.servicecall.core.Context;
import bma.servicecall.core.Request;
import bma.servicecall.core.Value;
import bma.servicecall.core.ValueMap;
import bma.servicecall.core.ValueMapWalker;

public class MessageWriter extends OutputStream {
	private OutputStream w;
	private byte[] hbuf = new byte[SocketCoreConst.HEADER_SIZE];
	private ByteArrayOutputStream buf;

	public MessageWriter(OutputStream w) {
		super();
		this.w = w;
	}

	@Override
	public void close() throws IOException {
		super.close();
		if (this.buf != null) {
			this.buf.close();
			this.buf = null;
		}
	}

	@Override
	public void write(int b) throws IOException {
		this.w.write(b);
	}

	@Override
	public void write(byte[] b, int off, int len) throws IOException {
		this.w.write(b, off, len);
	}

	public void writeHeader(int mt, int sz) throws IOException {
		this.hbuf[0] = (byte) (mt & 0xFF);
		this.hbuf[1] = (byte) (sz >> 16 & 0xFF);
		this.hbuf[2] = (byte) (sz >> 8 & 0xFF);
		this.hbuf[3] = (byte) (sz & 0xFF);
		this.write(this.hbuf);
	}

	public void writeEnd() throws IOException {
		this.write(Coder.endData);
	}

	public void writeMessageId(int mid) throws IOException {
		this.writeHeader(SocketCoreConst.MT_MESSAGE_ID, 4);
		Coder.encodeFixInt32(this, mid);
	}

	public void writeFlag() throws IOException {
		this.write(Coder.flagData);
	}

	protected ByteArrayOutputStream sbuf() {
		if (this.buf == null) {
			this.buf = new ByteArrayOutputStream();
		}
		this.buf.reset();
		return this.buf;
	}

	public void writeAddress(String s, String m) throws IOException {
		ByteArrayOutputStream buf = this.sbuf();
		int n = 0;
		n += Coder.encodeLenString(buf, s);
		n += Coder.encodeLenString(buf, m);
		this.writeHeader(SocketCoreConst.MT_ADDRESS, n);
		this.write(buf.toByteArray());
	}

	public void writeData(String name, Value val) throws IOException {
		ByteArrayOutputStream buf = this.sbuf();
		int n = 0;
		n += Coder.encodeLenString(buf, name);
		n += Coder.encodeVar(buf, val);
		this.writeHeader(SocketCoreConst.MT_DATA, n);
		this.write(buf.toByteArray());
	}

	public void writeContext(String name, Value val) throws IOException {
		ByteArrayOutputStream buf = this.sbuf();
		int n = 0;
		n += Coder.encodeLenString(buf, name);
		n += Coder.encodeVar(buf, val);
		this.writeHeader(SocketCoreConst.MT_CONTEXT, n);
		this.write(buf.toByteArray());
	}

	public void writeAnswer(int st, String msg) throws IOException {
		ByteArrayOutputStream buf = this.sbuf();
		int n = 0;
		n += Coder.encodeInt32(buf, st);
		n += Coder.encodeLenString(buf, msg);
		this.writeHeader(SocketCoreConst.MT_ANSWER, n);
		this.write(buf.toByteArray());
	}

	public void sendRequest(int mid, String s, String m, Request req,
			Context ctx) throws IOException {
		this.writeMessageId(mid);
		this.writeFlag();
		this.writeAddress(s, m);
		if (req != null) {
			req.walk(new ValueMapWalker() {

				@Override
				public boolean walk(String k, Value v) {
					try {
						writeData(k, v);
					} catch (IOException e) {
						throw AppError.handle(e);
					}
					return false;
				}
			});
		}

		if (ctx != null) {
			ctx.walk(new ValueMapWalker() {

				@Override
				public boolean walk(String k, Value v) {
					try {
						writeContext(k, v);
					} catch (IOException e) {
						throw AppError.handle(e);
					}
					return false;
				}
			});
		}
		this.writeEnd();
	}

	public void sendAnswer(int mid, Answer an) throws IOException {
		this.writeMessageId(mid);
		this.writeAnswer(an.getStatus(), an.getMessage());

		ValueMap rs = an.getResult();
		if (rs != null) {
			rs.walk(new ValueMapWalker() {

				@Override
				public boolean walk(String k, Value v) {
					try {
						writeData(k, v);
					} catch (IOException e) {
						throw AppError.handle(e);
					}
					return false;
				}
			});
		}

		ValueMap ctx = an.getContext();
		if (ctx != null) {
			ctx.walk(new ValueMapWalker() {

				@Override
				public boolean walk(String k, Value v) {
					try {
						writeContext(k, v);
					} catch (IOException e) {
						throw AppError.handle(e);
					}
					return false;
				}
			});
		}
		this.writeEnd();
	}
}
